src/sas/sascalc/file_converter/ascii2d_loader.py

```diff
@@ -0,0 +1,143 @@
+"""
+ASCII 2D Loader
+"""
+from sas.sascalc.dataloader.data_info import Data2D
+from sas.sascalc.file_converter.nxcansas_writer import NXcanSASWriter
+import numpy as np
+
+# ISIS 2D ASCII File Format
+# http://www.isis.stfc.ac.uk/instruments/loq/software/colette-ascii-file-format-descriptions9808.pdf
+# line: property
+# 0: File header
+# 1: q_x axis label and units
+# 2: q_y axis label and units
+# 3: Intensity axis label and units
+# 4: nUseRec - number of lines of user content following this line
+# 5 - 5+nUseRec: user content
+# Number of qx points
+# List of qx points
+# Number of qy points
+# List of qy points
+# numqx numqy scale
+
+class ASCII2DLoader(object):
+
+    def __init__(self, data_path):
+        """
+        :param data_path: The path to the file to load
+        """
+        self.data_path = data_path
+
+    def load(self):
+        """
+        Load the data from the file into a Data2D object
+
+        :return: A Data2D instance containing data from the file
+        :raises ValueError: Raises a ValueError if the file is incorrectly formatted
+        """
+        file_handle = open(self.data_path, 'r')
+        file_buffer = file_handle.read()
+        all_lines = file_buffer.splitlines()
+
+        # Load num_points line-by-line from lines into a numpy array, starting
+        # on line number start_line
+        def _load_points(lines, start_line, num_points):
+            qs = np.zeros(num_points)
+            n = start_line
+            filled = 0
+            while filled < num_points:
+                row = np.fromstring(lines[n], dtype=np.float32, sep=' ')
+                qs[filled:filled+len(row)] = row
+                filled += len(row)
+                n += 1
+            return n, qs
+
+        current_line = 4
+        try:
+            # Skip nUseRec lines
+            nUseRec = int(all_lines[current_line].strip()[0])
+            current_line += nUseRec + 1
+            # Read qx data
+            num_qs = int(all_lines[current_line].strip())
+            current_line += 1
+            current_line, qx = _load_points(all_lines, current_line, num_qs)
+
+            # Read qy data
+            num_qs = int(all_lines[current_line].strip())
+            current_line += 1
+            current_line, qy = _load_points(all_lines, current_line, num_qs)
+        except ValueError as e:
+            err_msg = "File incorrectly formatted.\n"
+            if str(e).find('broadcast') != -1:
+                err_msg += "Incorrect number of q data points provided. "
+                err_msg += "Expected {}.".format(num_qs)
```

Comment:
If a ValueError is raised on line 58, the error message is set on line 75, so num_qs is never used